### PR TITLE
Adding risk message validation++

### PR DESCRIPTION
**Pull request by cmcginley-splunk**
# Context
* As part of PEX-363, we wanted to expand integration testing to validate risk message content
* If invalid fields are referenced in the risk message, they will fail to render silently within ES

# Code changes
* Added new `RiskEvent` model for the events returned by ES
* Refined search criteria in`CorrelationSearch` to only search for events which match the appropriate search name
* Risk validation expanded to check:
  * Risk message contains no `$...$` literals (**now scans for the specific field replacement tokens rather than the `$...$` pattern)
  * Risk message in detection matches risk message in risk event (regex conversion)
  * Risk score matches detection
  * Analytic stories match detection
  * MITRE ATT&CK IDs match detection
* Added new `NotableEvent` model for the events returned by ES (will support additional notable validation, potentially added in the future)
* Removed a bug where we we were sleeping for 60s initially twice (should remove 20h from our cumulative compute time, about 30min on our total runtime across 40 instances)
* Refined cleanup logic
* Refactored `format_pbar_string` s.t. it uses the `start_time` instance attribute if none is provided explicitly
* Linted/formatted/typed:
  * `observable.py`
  * `detection_abstract.py`
  * `security_content_object_abstract.py`
  * `atomic.py`
  * `base_test_result.py`
  * `detection_tags.py`
* Added debug configuration where you can select your args; must supply path to `security_content` repo as an env var (SECURITY_CONTENT_PATH)

# New detection failures
~~I spot checked these, but did not do a deep dive on every single one; but I believe they are all legitimate validation issues~~
* ~~**59 detections** are failing this new validation (tested on v4.29.0)~~
  * ~~51 detections create risk events where the risk message contains a `$...$` literal -> represents a bad field substitution, likely because the referenced field doesn't exist in final SPL output~~
  * ~~8 detections have mismatches between the analytic stories listed in the detection and what's observed in the risk event -> spot checking shows this is mostly due to typos/casing in analytic story names in the detection which fails to link the actual analytic story to the risk event in ES~~
 
~~**NOTE**: this testing was performed locally, and some detections failed due to networking issues, likely due to my ISP bandwidth; so there may be more legitimate failures similar to the above not captured in this initial test~~

* **UPDATE 8/7/24**: 
  * Ran full package test on develop
  * Three detections failed risk message validation after refinement (tracked for resolution internally)
  * All false positives have been eliminated w/ the new method of scanning the risk_message for the field replacement tokens instead of the `$...$` pattern

# Testing
Will post some results from an SCA pipeline when that run completes

# Future work
* This PR also adds some commented out code which matches risk events to observables in the detection
  * This feature is still in progress as it does generate some false positives
  * That said, it did expose some legitimate issues (e.g. Attacker role observables creating risk events), so it will be a good addition once that validation logic has been refined to remove false positives.
* **Add validation which enforces the usage of field names w/o spaces; until this is enforced, risk message validation will not be able to validate risk messages leveraging such fields as the extraction pattern we use for pulling the field tokens to scan for assumes no spaces are present**
* **Add validation which checks for balanced `$` in `tags.message`; if a need for a `$` literal in messages is found, we can consider adding some kind of escaping to support it**

### CODE REVIEW

File: contentctl/objects/correlation_search.py
Comment (cmcginley-splunk) [inline contentctl/objects/correlation_search.py:37]: Will set this back to False after a final full test w/in an SCA pipeline

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py
Comment (patel-bhavin) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py]: Yes, contentctl validate fails today, when a production detection does not have tests: Verified by removing tests from production detection
image

File: contentctl/objects/correlation_search.py
Comment (patel-bhavin) [reply contentctl/objects/correlation_search.py:35]: @cmcginley-splunk : Should we set this back to false now that we have ran this a few times in SCA repo?

**Comment by cmcginley-splunk** [comment]
Some detections seem to be generating false positives (test failures that are not actual failures); e.g.: Windows Excessive Disabled Services Event

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py
Comment (cmcginley-splunk) [inline contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py:239]: Can this be removed?

File: contentctl/enrichments/attack_enrichment.py
Comment (cmcginley-splunk) [inline contentctl/enrichments/attack_enrichment.py:33]: do we want to return None when use_enrichment is false? This seems confusing, and it seems that we intentionally do not call this function when that setting is enabled in `DetectionTags.addAttackEnrichment`. I'm unclear why we have a separate invocation of this in `SSADetectionBuilder`. I'd vote to streamline the usage of this and have this raise if this flag is set and this function is called.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:168]: We have `file_path` as a required field; this check is irrelevant no? Pydantic validation precludesthe need for this check

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:245]: Is this field even used?

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:397]: can we remove this commented out null type check?

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:454]: This is an instance of some of the weird typing situations we run into w/ `Detection` vs `Detection_Abstract`. `self` is of type `Detection_Abstract` but `story.detections` is declared as a list of `Detection` objects. This part of the `model_post_init` should be moved to the `Detection` class; or, better yet, we should remove the `Detection_Abstract` class altogether IMO.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:487]: If it's possible in practice for this to return `None` that should be part of the type, otherwise the check on line 488 makes no sense.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:556]: can we remove any of this dead code?

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:513]: if search is a dict, what's the possible typing? is it `dict[str, str]`?

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/detection_abstract.py:585]: We are returning a list of `SecurityContentObject_Abstract`; we should be returning `Detection_Abstract`; perhaps `Detection_Abstract` should have an overwridden version of `create_filename_to_content_dict` to accomodate this?

File: contentctl/objects/abstract_security_content_objects/security_content_object_abstract.py
Comment (cmcginley-splunk) [inline contentctl/objects/abstract_security_content_objects/security_content_object_abstract.py:9]: Can we remove this `Config`? It's not used anywhere and does not seem to exist.

Additionally, what's happening here w/ the check against `TYPE_CHECKING`?

File: contentctl/objects/detection_tags.py
Comment (cmcginley-splunk) [inline contentctl/objects/detection_tags.py:110]: I believe this should be a list of of `dict[str, str]` yes? Perhaps we should make a model for `Mapping` since it has consistent key structure?

File: contentctl/objects/lookup.py
Comment (cmcginley-splunk) [inline contentctl/objects/lookup.py:29]: We do we parse `fields_list` as a CSV string and not a list of strings?

File: contentctl/objects/unit_test.py
Comment (cmcginley-splunk) [inline contentctl/objects/unit_test.py:12]: Do we need this commented out code?

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:168]: Agreed, I think was a relic from a time when file_path was not a required_field (and that functionality doesn't even work right now, since it raises an exception).  I have removed that check and now just returned the parent name.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:481]: Yup, that's correct! I will make this change as well.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:550]: Yes, I have removed all that dead code.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:507]: The searches that are dicts are sigma searches, so the typing for those should be dict[str,Any]. I will update that typing, and the typing for the search field, to reflect that.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:579]: This function is actually not used anymore, so it has been removed.

File: contentctl/objects/abstract_security_content_objects/security_content_object_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/security_content_object_abstract.py:9]: Yes, this has now been removed.

File: contentctl/objects/baseline_tags.py
Comment (pyth0n1c) [reply contentctl/objects/baseline_tags.py:21]: This typing is weird, but intentional.  In the baseline files, we define the related searches initially as a list of strings.  Then, we construct the Baseline objects.  However, at this time, the related Detection objects do not exist!

Then, later, we construct the detections and iterate over the Baselines and convert the list of detection name strings into a list of Detection OBJECTS. You can see the code for this here:
https://github.com/splunk/contentctl/blob/96b813fa220ffef6349487421687b99404f647a3/contentctl/objects/abstract_security_content_objects/detection_abstract.py#L401-L415

You will notice that because Detections are constructed after Baselines, Macros, Lookups, etc, they don't have this weird double-typing:  https://github.com/splunk/contentctl/blob/96b813fa220ffef6349487421687b99404f647a3/contentctl/objects/abstract_security_content_objects/detection_abstract.py#L199

I would love to do this a different way, but can't think of it - I am very open to suggestions.

File: contentctl/objects/detection_tags.py
Comment (pyth0n1c) [reply contentctl/objects/detection_tags.py:107]: ABSOLUTELY. Anything we can do to move away from `dicts` is a huge win in my book.
I will leave this one open as I haven't made that change yet.

File: contentctl/objects/detection_tags.py
Comment (pyth0n1c) [reply contentctl/objects/detection_tags.py:240]: I have changed the typing for DirectorOutputDto.atomic_tests to be list[AtomicTest] instead. Now, atomic_tests will be an empty list if we have not read any from the repo. I have verified that this change still works as expected whether enrichment is enabled/disabled.

File: contentctl/objects/lookup.py
Comment (pyth0n1c) [reply contentctl/objects/lookup.py:29]: This is probably somewhat-legacy code/typing.  In the YMLs, fields is a comma-separated string. We should likley convert all the YMLs to use a list[str].
Additionally, Lookups are actually two different kinds of lookups - CSVLookup and KVStore Lookup. These have have different requirements. When we make the above change, we should consider breaking this into two different classes with Lookup as an Abstract Base Class.

I will leave this comment open for discussion

File: contentctl/objects/unit_test.py
Comment (pyth0n1c) [reply contentctl/objects/unit_test.py:12]: Yes, they have been removed.

File: contentctl/objects/lookup.py
Comment (cmcginley-splunk) [reply contentctl/objects/lookup.py:29]: How do we want to track small quality of life improvements like this? GitHub issues? Internally on Jira?

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py
Comment (cmcginley-splunk) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py]: Removed this where? I don't see the change on this branch

File: contentctl/objects/lookup.py
Comment (cmcginley-splunk) [reply contentctl/objects/lookup.py:29]: See https://github.com/splunk/contentctl/issues/220

File: contentctl/objects/detection_tags.py
Comment (cmcginley-splunk) [reply contentctl/objects/detection_tags.py:107]: See https://github.com/splunk/contentctl/issues/221

File: contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py
Comment (pyth0n1c) [reply contentctl/actions/detection_testing/infrastructures/DetectionTestingInfrastructure.py]: Yes, this has been removed.

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (pyth0n1c) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:244]: This is legacy code which is related to DocGen Functionality, which is no longer exposed. I have raised an internal issue with the team to see if DocGen can/should be removed. If so, that will be a much larger PR but should be a nice cleanup to numerous files :) 
As such, I am going to leave this comment unresolved, but no relevant changes should be part of THIS PR.

File: contentctl/objects/unit_test.py
Comment (cmcginley-splunk) [inline contentctl/objects/unit_test.py:4]: What're we doing here w/ these conditional imports? Esp. bc we import them again a few lines later

File: contentctl/objects/baseline_tags.py
Comment (cmcginley-splunk) [reply contentctl/objects/baseline_tags.py:21]: Understood! Will investigate later as part of #223

File: contentctl/objects/abstract_security_content_objects/detection_abstract.py
Comment (cmcginley-splunk) [reply contentctl/objects/abstract_security_content_objects/detection_abstract.py:445]: Resolving for now; I think theres a larger design convo to be had about `Detection` and `Detection_Abstract`

File: contentctl/objects/unit_test.py
Comment (pyth0n1c) [reply contentctl/objects/unit_test.py:4]: Good catch. These should not be double-imported.
The conditional imports are because many of our objects are structured...interesting and _intentionally_ do circular imports.  For example, a Story includes a list of Detections and a Detection includes a list of Stories.

For typing reasons, Story needs to import Detection and Detection needs to import Story - but Python/Pydantic throw parse exceptions if you don't import them (as expected). However, a regular import throws a _different_ runtime Pydantic exception about models not being fully defined.

Putting them imports in an if `TYPE_CHECKING` block satisfies the type checker and avoids the runtime errors mentioned above. It's a bit of a mess - we may be able to do this better by _explicitly_ calling `PydanticObject.model_rebuild` after all objects have been imported, but I haven't been able to resolve that yet.

**Review by pyth0n1c — APPROVED** [review]
Casey and I have discussed this and made some improvements based on the suggested feedback. We have also opened some issues for ongoing, related work.
This is approved - we will wait on the outcome of some internal testing before actually merging it in.

**Review by pyth0n1c — APPROVED** [review]
I bumped the version, which dismissed my previous approval. Just approving again

**Review by patel-bhavin — APPROVED** [review]

**Pull request merged**
